src/features/class-block-audit-command.php

```diff
@@ -0,0 +1,275 @@
+						 *
+						 * @param string $example_url The example URL. Defaults to the permalink of the post where the block type was first seen.
+						 */
+						$example_url = apply_filters( 'alley_block_audit_block_type_example_url', get_permalink( $post ) );
+
+						$out[ $block_name ] = [
+							'Block Name'  => $block_name,
+							'Count'       => 0,
+							'Example URL' => $example_url,
+							'Post Types'  => [],
+							'Details'     => [],
+						];
+					}
+
+					$out[ $block_name ]['Count']++;
+
+					if ( ! in_array( $post->post_type, $out[ $block_name ]['Post Types'], true ) ) {
+						$out[ $block_name ]['Post Types'][] = $post->post_type;
+					}
+
+					$out[ $block_name ]['Details'] = $this->with_details(
+						$out[ $block_name ]['Details'], // @phpstan-ignore-line
+						$block, // @phpstan-ignore-line
+					);
+				}
+			},
+		);
+
+		if ( count( $out ) === 0 ) {
+			\WP_CLI::warning( 'No results. Run again with the --rewind flag to reset the cursor.' );
+			return;
+		}
+
+		// Sort by block name.
+		ksort( $out );
+		$first = reset( $out );
+
+		foreach ( $out as &$values ) {
+			// Sort details.
+			ksort( $values['Details'] );
+
+			// Leave details empty if there are none.
+			if ( [] === $values['Details'] ) {
+				$values['Details'] = '';
+			}
+		}
+
+		$format = get_flag_value( $assoc_args, 'format' );
+
+		format_items(
+			is_string( $format ) && $format ? $format : 'table',
+			$out,
+			array_keys( $first ),
+		);
+	}
+
+	/**
+	 * Add to details about blocks of a certain type.
+	 *
+	 * For example:
+	 *
+	 * - Keep track of which heading levels are in use.
+	 * - Keep track of how many recirculation modules use custom post titles.
+	 *
+	 * @phpstan-param array<string, array<string, mixed>> $details
+	 * @phpstan-param array{blockName: string, attrs: array<string, mixed>, innerHTML: string} $block
+	 * @phpstan-return array<string, mixed>
+	 *
+	 * @param array $details Accumulated details about blocks of this type so far.
+	 * @param array $block   Block being audited.
+	 * @return array Updated details.
+	 */
+	private function with_details( array $details, array $block ): array {
+		$html  = new \WP_HTML_Tag_Processor( $block['innerHTML'] );
+		$attrs = $block['attrs'];
+
+		// Automatically track standard alignment attribute.
+		if ( isset( $attrs['align'] ) && is_string( $attrs['align'] ) ) {
+			$details['align'][ $attrs['align'] ] ??= 0;
+			$details['align'][ $attrs['align'] ]++;
+		}
+
+		switch ( $block['blockName'] ) {
+			case 'core/embed':
+				if ( ! empty( $attrs['providerNameSlug'] ) ) {
+					$details['providerNameSlug'][ $attrs['providerNameSlug'] ] ??= 0;
+					$details['providerNameSlug'][ $attrs['providerNameSlug'] ]++;
+				}
+				break;
+
+			case 'core/heading':
+				while ( $html->next_tag() ) {
+					$tag = $html->get_tag();
+
+					if ( is_string( $tag ) && preg_match( '/^H\d$/', $tag ) ) {
+						$details[ $tag ] ??= 0;
+						$details[ $tag ]++;
+					}
+				}
+				break;
+
```

Comment:
should we add an action or filter here so that attributes on other blocks can be handled?

Reply:
Yeah, there definitely should be some way to add custom details, but I put it off because of the performance implications. Like, on the site we launched last summer, with 100,000+ published posts, that might add up to millions of `apply_filters()` calls. So I wanted to think more about how to optimize it. Maybe it'd be negligible in the end, I dunno.

Reply:
OK, it was gnawing at me, so I put something in place. Thanks for the nudge! 😄